johnzon-jsonb/src/main/java/org/apache/johnzon/jsonb/JsonbPolymorphismHandler.java

```diff
@@ -0,0 +1,155 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.johnzon.jsonb;
+
+import org.apache.johnzon.mapper.access.Meta;
+
+import jakarta.json.JsonObject;
+import jakarta.json.JsonString;
+import jakarta.json.JsonValue;
+import jakarta.json.bind.JsonbException;
+import jakarta.json.bind.annotation.JsonbSubtype;
+import jakarta.json.bind.annotation.JsonbTypeInfo;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class JsonbPolymorphismHandler {
+    public boolean hasPolymorphism(Class<?> clazz) {
+        return clazz.isAnnotationPresent(JsonbTypeInfo.class) || !getParentClassesWithTypeInfo(clazz).isEmpty();
+    }
+
+    public List<Map.Entry<String, String>> getPolymorphismPropertiesToSerialize(Class<?> clazz, Collection<String> otherProperties) {
+        List<Map.Entry<String, String>> entries = new ArrayList<>();
+
+        Class<?> current = clazz;
+        while (current != null) {
+            validateJsonbTypeInfo(current);
+            validateOnlyOneParentWithTypeInfo(current);
+
+            JsonbTypeInfo typeInfo = Meta.getDirectAnnotation(current, JsonbTypeInfo.class);
+            if (typeInfo != null) {
+                if (otherProperties.contains(typeInfo.key())) {
+                    throw new JsonbException("JsonbTypeInfo key '" + typeInfo.key() + "' collides with other properties in json");
+                }
+
+                if (entries.stream().anyMatch(entry -> Objects.equals(entry.getKey(), typeInfo.key()))) {
+                    throw new JsonbException("JsonbTypeInfo key '" + typeInfo.key() + "' found more than once in type hierarchy of " + clazz.getName());
+                }
+
+                String bestMatchingAlias = null;
+                for (JsonbSubtype subtype : typeInfo.value()) {
+                    if (subtype.type().isAssignableFrom(clazz)) {
+                        bestMatchingAlias = subtype.alias();
+
+                        if (clazz == subtype.type()) { // Exact match found, no need to continue further
+                            break;
+                        }
+                    }
+                }
+
+                if (bestMatchingAlias != null) {
+                    entries.add(0, Map.entry(typeInfo.key(), bestMatchingAlias));
+                }
+            }
+
+            List<Class<?>> parentClassesWithTypeInfo = getParentClassesWithTypeInfo(current);
+            current = parentClassesWithTypeInfo.isEmpty() ? null : parentClassesWithTypeInfo.get(0);
+        }
+
+        return entries;
+    }
+
+    public Class<?> getTypeToDeserialize(JsonObject jsonObject, Class<?> clazz) {
+        validateJsonbTypeInfo(clazz);
```

Comment:
this should be done one per class somehow since it is a runtime class (compared to getPolymorphismPropertiesToSerialize which is a classmapping init computation only) so maybe use ClassMapping to do the validation only one, can likely become a `Supplier<Class<?>>` using the supplier "pre" part to do the validation and `supplier#get` to impl the runtime only (current line 91+).